### DISCUSSION TITLE
Integration with Vim

### DISCUSSION THREAD

**Discussion by mickael-menu** [zk-org/zk#11] [General]
:warning: UPDATE: See https://github.com/mickael-menu/zk-nvim

@megalithic started this discussion in https://github.com/mickael-menu/zk/issues/10#issuecomment-799666109
> I'm looking at ways to make link-following easier; since this is tool agnostic for editing/etc; and you seem to be a (n)vim user; do you recommend something like vim-wiki/vimwiki for handling the link expansion and creation?
>
> Basically, I think I'm looking to utilize `zk` for management of zettels, as well as it's journaling feature for more long-form day-to-day work notes.

**Comment by megalithic** [comment]
Brilliant move! Thanks for wanting to chat through this. I've long been following @evantravers' setup for note-taking, in all it's various permutations. I'm still discovering what works best for me for my day to day. 

I fear that I might be trying to force one tool to handle multiple goals.

For instance:

1. keeping up with things related to a project I'm managing at work
1. keeping up with any long-form things like a blog post
1. keeping up single thought notes around code things, dotfiles things, house projects, Bible related things, book reading quotes, etc
1. linking all of these in a way that lets me easily cross-ref/relate

I use Drafts.app on mac and iOS for quick entry of things to then use an action @evantravers wrote to convert the draft (original draft is archived in the drafts app itself) to a zettel/obsidian friendly .md file to then use neovim to edit further.

I will also just go to town in prose mode via neovim to just type out blog posts and ideas, etc.

**Reply by mickael-menu** [reply]
Some tools might definitely be more convenient for certain use cases. I've been a happy user of [Diarly](https://diarly.app) and [Things](https://culturedcode.com/things/) before, but I found that keeping everything in the same simple plain text system was helpful to cross-reference things, enable Git versioning and scripting.

Here are some stuff I keep in my notebook:

* Daily logs including tasks and pomodoro counts. When I face a particular problem, I might create a dedicated note and link it from the daily note.
* Troubleshooting notes or useful code snippets.
* Raw reading notes.
* Recipes.
* Proper "evergreen notes"/zettels on subjects I want to really understand.

I'm still experimenting things to find the most convenient organization, so far I have my daily notes in `log/`, which allows me to easily filter out the "noise" with `zk list -x log`. For the rest, I keep everything at the root directory and use tags. I added tags in `zk` only very recently, you could also filter them out with `zk list -t "NOT log"`.

So far I've "marketed" `zk` as a Zettelkasten tool to emphasize the linking, but it's actually pretty usage agnostic. Mine might actually be closer to a personal wiki sometimes!

**Comment by mickael-menu** [comment]
> I'm looking at ways to make link-following easier; since this is tool agnostic for editing/etc; and you seem to be a (n)vim user; do you recommend something like vim-wiki/vimwiki for handling the link expansion and creation?

That's right, at the moment I've been using [vimwiki](https://github.com/vimwiki/vimwiki) coupled with [vim-zettel](https://github.com/michal-h21/vim-zettel) and [notational-fzf-vim](https://github.com/alok/notational-fzf-vim). This works great for navigating through my notes and creating links, but a better integration with `zk` – especially to create new notes – would be nice.

I'm thinking of building a dedicated Vim plugin eventually, especially since I don't really need the vimwiki part which is large. With a better `zk` integration, we could have stuff like side buffers showing backlinks or unlinked mentions. I have no experience with Vimscript though, but I saw that Go could be used, at least with Neovim.

**Reply by megalithic** [reply]
I love this idea! I was just thinking through a lightweight lua-based plugin to sit on top of `zk`'s core. I love the idea of utilizing some of neovim's fancier features (virtual text, popups, those sidebars, and other things) to create an obsidian-like experience, just shy of the fancy viz.

One neat thing might be to use the virtual text to place a related/ref count next to a specific tag or zettel/page link to get a quick glance at how many inter-linked things refer to that specific tag; just a simple example. 

Anyway, I'll be watching this project even closer, especially as I use it more and get acclimated to it and finding my own workflow.

**Reply by evantravers** [reply]
I am pretty sure that the lightweight lua plugin is the way.

As you and I were texting earlier Seth… the key features:

- links. This alone would replace vimwiki largely. 
- note creation
- Search

@mickael-menu has done the heavy lifting… I feel like a couple aliases and examples and we are off to the races.

**Reply by megalithic** [reply]
I have a grand news; i've tossed together a super quick lua neovim plugin, untested. just wanted to get thoughts and a structure down to start tinkering.

I'd love to add you gents as collaborators if interested.
https://github.com/megalithic/zk.nvim

**Reply by megalithic** [reply]
I've added a few things; just a basic skeleton to creating commands: https://github.com/megalithic/zk.nvim/compare/feat/main/impl_first_command

Not done, just wanted to get this out there.

**Reply by mickael-menu** [reply]
That's awesome! Thanks for tackling this, you seem quite knowledgeable about Neovim. 👍 I agree that the lightweight "UX" wrapper feels like the right approach.

IMHO, one of the most useful feature from `vim-zettel` is the creation of links with auto-completion when typing `[[`. `zk list --interactive` should be able to provide the UX to search through the notes. It would be nice to be able to create a new note from this screen as well. It's possible with `zk edit -i` and `Ctrl-N`, but not sure this can be easily integrated with Vim.

Later on for backlinks and such, I think we can have a highly customizable approach by specifying any number of "dynamic" buffers generated with `zk list`. Here are a few examples:

* `--link-to X` – backlinks
* `--related X` – notes which have one "connection" in common with X but not linked to it
    * I plan to refine `related` to find notes which have similar themes using NLP
* `--mention X --no-link-to X` – notes mentioning X's title or aliases, but not linked to it
* `--sort modified --limit 10` – 10 notes recently modified

By the way these two notes might be relevant for the wrapper:

* [Call `zk` from other programs](https://github.com/mickael-menu/zk/blob/main/docs/external-call.md)
* [Send notes for processing by other programs](https://github.com/mickael-menu/zk/blob/main/docs/external-processing.md)

**Comment by megalithic** [comment]
@mickael-menu and @evantravers 

I've implemented the first couple of commands.. a way to install `zk` if golang is installed on the user's system; and a way to create a new note with an optional title argument. The issue is, right now I don't check for the `.zk` folder as a root marker. I think this is kind of the rub is that you must be in the cwd of your initialized notebook. 

While not a _huge_ deal; it will just require, maybe some basic configuration for the plugin, basically defining a notebook dir to use as the base path for all commands to be executed.

Thoughts?

https://github.com/megalithic/zk.nvim/tree/feat/main/impl_first_command

**Reply by mickael-menu** [reply]
* `zk` works in any subdirectory of a notebook, `zk list` will display the paths relative to the working directory in this case.
    * However, sometimes it's useful to make sure the commands are ran from the root of the notebook, in case we use specific predefined paths (e.g. `-x log` to ignore the daily logs in the search).
    * I'm using [vim-rooter](https://github.com/airblade/vim-rooter) which conveniently moves the wd to the root of a file's Git repository if there's any. Since I also use Git in my notebook, that's enough. But maybe we could have something similar in `zk.nvim` to auto-cwd to the root of the current notebook.
* I think it makes sense to define a default notebook in `zk.nvim`'s config, to use with commands when not already in a notebook.
    * The cwd notebook should take precedence over the default one, otherwise this limits the plugin to a single notebook.
    * Would it help if `zk` returned a particular exit code meaning "not in a notebook"? This way you could first attempt the command in cwd, and fallback on the notebook in the config.

On a side note, by default `zk new` and `zk edit` will open your editor, but it might be more convenient to get the file paths instead to let the control to `zk.nvim`. For example that could be useful to say "create a new note and open it in a vertical split". This might help:

* use `zk new --print-path` to print the path to stdout instead of opening the editor
* use `zk list --format path [filtering criteria]` to get the list of filepaths matching the filters (it works with `--interactive` too, for fzf)

**Reply by megalithic** [reply]
Brilliant!

**Reply by megalithic** [reply]
Also, @mickael-menu please let me know if you'd prefer to not really have involvement in the plugin dev; in other words, if you don't care to be on the hook for PRs and such :) I'm good to move present PR/feature-branch to `main` of the plugin. 

You answered some of the main things I needed to know to keep moving. :)

**Reply by mickael-menu** [reply]
No I'm really enthusiastic about this whole project! I don't know a thing about Lua but I will follow closely your PRs. You don't need to wait for my reviews before merging though. I'm watching your repo so I will see all the issues and PRs.

Let me know if you need me to add the "not in a notebook" exit status code.

**Reply by megalithic** [reply]
@mickael-menu i've started some discussions back here: https://github.com/megalithic/zk.nvim/discussions/9

Thinking maybe we could move the vim/nvim related discussions there, to keep your repo focused on the core functionality?

Anyway, this is super exciting and really looking forward to making use of all of this even further. I've not fully realized my workflow yet; just jumped into this plugin while learning your lib, too.

**Reply by mickael-menu** [reply]
Sounds good 👌

**Comment by megalithic** [comment]
Just wanted to give a quick update on where we're at. I just merged the ability
to generate a new note from word under cursor (or visual selection). See the
[README](https://github.com/megalithic/zk.nvim#generate-a-new-note-and-inline-link) for more information on how to use it and how it works.

**Reply by mickael-menu** [reply]
@megalithic Sounds great! I'm still on nvim 0.4 but will upgrade eventually.

**Reply by mickael-menu** [reply]
By the way I will add a way to format a note link in `zk` directly, maybe you could use it in `zk.nvim` to have a single place where the user setups the link format. It's useful to have in `zk` for the LSP server and also later for generating stuff like a table of contents.

**Reply by megalithic** [reply]
That'd be fantastic @mickael-menu! I have some thoughts around this feature that
aren't _exactly_ what I want, but close. I don't quite have an elegant way of
where to generate the auto-created notes/links; presently they go to the
notebook you define in the function for creating the note link; that would
normally be okay, however, sometimes you might be in your wiki folder and want
to createa  new link IN that folder, you'd have to end up having multiple
bindings for each of your notebooks to create the link in; so i was wondering if
it'd be smart to try and derive the current notebook you're in, basically take
the `cwd` and then the `default_notebook_dir` you set and take that "sub dir" or
"sub notebook" as the current notebook the linked/new note should be in. I hope
that makes sense.

**Reply by mickael-menu** [reply]
I'm not sure I understand what you mean. Do you have some examples with concrete paths?

**Reply by megalithic** [reply]
@mickael-menu .. i thought it best to just create a loom to hopefully illustrate
the issue. please let me know if this makes sense. to be clear though, this _is
not_ an issue with zk, it is just the way i've implemented (or, not implemented)
the create_link functionality within zk.nvim. 

https://www.loom.com/share/7506d8ab349644c0907a0203fe1850b8

**Reply by mickael-menu** [reply]
Thanks for the screencast, it was definitely helpful!

Just so we're on the same page for the terminology, I don't use the term sub or nested notebook. With `zk` there are just notebooks (directories with `.zk` marker) and (potentially nested) subdirectories inside a given notebook. So in your example `wiki` would be a directory in your `_notes` notebook.

For the issue you raised, it's a good question and I'm not sure there's a definite answer except a matter of taste / usage.

`zk` is working relative to the current directory if it is part of a notebook. If none is found, `zk` considers that the current working directory is `ZK_NOTEBOOK_DIR`. So given a notebook `/notebook` and a directory `/notebook/wiki`, you can create a new note in `wiki` if:

* `pwd` gives `/notebook` and you call `zk new wiki`
* `pwd` gives `/notebook/wiki` and you call `zk new`
* `pwd` gives `/outside` and you call `ZK_NOTEBOOK_DIR=/notebook zk new wiki`
* `pwd` gives either `/notebook` or `/notebook/wiki` and you call `zk` with an absolute path: `zk new /notebook/wiki` (this will be relevant later)

I think this behavior can be helpful with Neovim, if you call `zk` from the folder returned by `:pwd`. For example if Neovim is opened in `/notebook` and you have the note `/notebook/wiki/foo.md` opened. Creating a new note from there will by default put it in `/notebook` because this is the current working directory. Now if you know you want to create notes in `/notebook/wiki` instead, you can open Neovim in `/notebook/wiki` directly or use `:cd wiki`, before creating the new note.

This would be a good basic behavior IMHO, but maybe not so practical if you edit notes across different directories, or if you have an `autocd` Neovim plugin that moves the `wd` to the root of the current Git repository. In this case, it could be useful if `zk.nvim` was providing two different ways to create a note:

1. Create a note relative to `:pwd` (with an optional `directory` parameter)
    * It would be nice to have an option for an interactive prompt to enter manually the `directory` parameter when creating a new note.
2. Create a "local" note, relative to the currently opened file
    * As long as `zk` is run from inside a notebook, you can actually use an absolute path to one of the notebook directory: `zk new /notebook/wiki`. I guess that could be convenient to derive the folder of the current file with Neovim and call `zk` with `zk new %:p:h`.
    * So basically the same as 1. but auto-setting `directory` to `%:p:h`.

**Reply by mickael-menu** [reply]
While I think two "quick" keybindings make sense depending on your usage, the interactive `directory` prompt could be used for a single interactive action. When triggered the user could enter:

* blank/nothing, to create a note in `:pwd` (`zk new`)
* a path to create a note in a subdirectory of `:pwd`, so if in `/notebook`, entering `wiki` would create the note in `/notebook/wiki` (`zk new wiki`)
* `.` (or something else?) to create a note relative to the currently opened file (`zk new %:p:h`)

In any case, we'll probably need to refine all this after some daily practice ;)

**Reply by mickael-menu** [reply]
@megalithic I just merged in the internal link formatter in `zk`: https://github.com/mickael-menu/zk/pull/32
The link syntax is fully customizable in the user config and more importantly, the path is relative to the working directory. I also added a `--working-dir` option to change the working directory used by `zk`, with this you can create a link relative to a note, for example:

```sh
$ zk list --quiet --format "{{link}}" --working-dir `pwd`/docs `pwd`/docs/automation.md
[Automating frequent tasks](automation.md)
```

I used `pwd` because it's easier to work with absolute paths in this case, as argument.

The LSP server uses this new link formatter to follow the user preferences, and is only triggered with double brackets `[[`.

Another use case, generating a valid Markdown list from filtering options:

```sh
$ zk list --match "link format" --format "* {{link}}"
* [Note formats](note-format.md)
* [Template context when formatting a note](template-format.md)
* [Changelog](../CHANGELOG.md)
* [Getting started with zk](getting-started.md)
* [Template syntax](template.md)
* [Send notes for processing by other programs](external-processing.md)
* [Notebook housekeeping](notebook-housekeeping.md)
* [Command aliases](config-alias.md)
* [Note configuration](config-note.md)
```

Here's a pre-built binary for macOS: [zk-v0.3.0-9-g2bb4cbd-macos-x86_64.zip](https://github.com/mickael-menu/zk/files/6331475/zk-v0.3.0-9-g2bb4cbd-macos-x86_64.zip)

**Comment by mhanberg** [comment]
This is what I'm using for navigating notes, tags, links in neovim.

It basically shells out `zk list -f json` in various forms to fzf.vim. I found I needed to basically remake `zk edit -i` to be able to do things like "`ctrl-v` to open note in vertical split" and the "`ctrl-n` to create a new note with query as title" easily.

https://github.com/mhanberg/.dotfiles/blob/main/config/nvim/plugin/zk.lua

```lua
local nnoremap = require("motch.utils").nnoremap
local _ = require("underscore")

vim.cmd([[command! ZkDaily :term zk daily]])

local colors = require("ansicolors")

local zk = {}

zk.note_picker = function(source)
  local notes = vim.fn.system(source)

  if vim.fn.trim(notes) == "" then
    notes = "[]"
  end

  notes = vim.fn.json_decode(notes)

  notes = _.map(notes, function(n)
    local tags = _.join(
      _.map(n.tags, function(t)
        return "#" .. t
      end),
      " "
    )

    return colors.noReset("%{bright}%{yellow}")
      .. n.title
      .. colors("%{reset}")
      .. " "
      .. ":"
      .. " "
      .. tags
      .. " "
      .. ":"
      .. " "
      .. n.absPath
  end)

  local actions = {
    ["enter"] = function(selected)
      vim.fn.execute("edit " .. selected)
    end,
    ["ctrl-v"] = function(selected)
      vim.fn.execute("vsplit " .. selected)
    end,
    ["ctrl-x"] = function(selected)
      vim.fn.execute("split " .. selected)
    end,
    ["ctrl-n"] = function(_, query)
      require("lspconfig/configs").zk.new({ title = query })
    end,
  }

  FZF({
    source = notes,
    sinklist = function(selected)
      local query = selected[1]
      local action = actions[selected[2]]
      local parts = vim.fn.split(selected[3], ":")

      action(parts[#parts], query)
    end,
    options = {
      "--expect",
      "enter,ctrl-v,ctrl-x,ctrl-n",
      "--header",
      colors("%{blue}CTRL-N: create a note with the query as title"),
      "--print-query",
      "--ansi",
      "--delimiter",
      ":",
      "--preview",
      "bat {-1}",
      "--nth",
      "1..2",
      "--prompt",
      "Notes> ",
    },
    window = { width = 0.9, height = 0.6, yoffset = 0, highlight = "Normal" },
  })
end

zk.list = function()
  zk.note_picker("zk list --quiet -f json --sort=created")
end

zk.tags = function()
  local tags = vim.fn.system("zk tag list --quiet -f json")

  tags = _.map(vim.fn.json_decode(tags), function(t)
    return t.name
  end)

  FZF({
    source = tags,
    sink = function(tag)
      zk.note_picker(string.format("zk list --quiet -f json --sort=created --tag %s", tag))
    end,
    options = { "--ansi", "--prompt", "Tags> " },
    window = { width = 0.9, height = 0.6, yoffset = 0, highlight = "Normal" },
  })
end

zk.backlinks = function()
  zk.note_picker(string.format("zk list --quiet -f json --sort=created --link-to %s", vim.fn.expand("%")))
end

zk.links = function()
  zk.note_picker(string.format("zk list --quiet -f json --sort=created --linked-by %s", vim.fn.expand("%")))
end

_G.motch.zk = zk

local dir = require("lspconfig").util.root_pattern(".zk")(vim.fn.getcwd())
if type(dir) == "string" then
  nnoremap("<C-p>", ":lua motch.zk.list()<cr>")
  nnoremap("<space>zt", ":lua motch.zk.tags()<cr>")
  nnoremap("<space>zb", ":lua motch.zk.backlinks()<cr>")
  nnoremap("<space>zl", ":lua motch.zk.links()<cr>")
  nnoremap("<space>zd", ":ZkDaily<cr>")
end
```

Note: The global `FZF` is just a lua wrapper of the vim fzf function to make it possible to use lua anonymous functions as the callback values. You can see the definition [here](https://github.com/mhanberg/.dotfiles/blob/main/config/nvim/init.lua#L16..L25).

**Reply by mickael-menu** [reply]
Amazing! I'll try it out.

Is there anything missing in `zk` that could help you out with such config?

For example, it would be nice to expose the [`fzf` config](https://github.com/mickael-menu/zk/blob/main/docs/tool-fzf.md) so that you could reuse `zk`'s `fzf-preview` and `fzf-line`. That could mean dynamic settings depending on the current notebook:

```sh
$ zk config get tool.fzf-preview
zk list --quiet --format full --limit 1 {-1}
```

And to render a template:

```sh
$ zk config render tool.fzf-line "{'title': 'A note', ...}"
%{yellow}A note %{black}Note content...
```

With a custom [`Style`](https://github.com/mickael-menu/zk/blob/439c1b1a6964f3ebd9237d40981cb17cc7a70d56/internal/core/style.go#L68) formatter, the `zk` style template helper (e.g. `{{style "yellow" title}}`) would be automatically converted into `fzf` color instructions.

**Reply by vod3** [reply]
it would be amazing to have something  similar for telescope!

**Reply by mhanberg** [reply]
> Amazing! I'll try it out.

Thanks!

> Is there anything missing in `zk` that could help you out with such config?

It would be cool to get the `fzf` options straight out of `zk`, but it does sound like a pretty specific use case. One thing I couldn't mimic was using the invisible (null?) character as the delimiter. Vim (or Lua?) was giving me problems, so I used a `:`, and will just try to not use colons in my note titles haha.

I think that my config will become easier and more powerful once there are more LSP commands in `zk`. Currently i run `local notes = vim.fn.json_decode(vim.fn.system(zk list --quiet -f json --sort created))`, but it would probably be faster to run `local notes = require("lspconfig/configs").zk.list({sort = "created"})`

**Reply by megalithic** [reply]
If you’re not streaming or posting on YouTube yet, you should be!!! I’d personally love to see your usage of this and for elixir too. :) thanks @mhanberg !!

**Reply by mickael-menu** [reply]
> I think that my config will become easier and more powerful once there are more LSP commands in `zk`. Currently i run `local notes = vim.fn.json_decode(vim.fn.system(zk list --quiet -f json --sort created))`, but it would probably be faster to run `local notes = require("lspconfig/configs").zk.list({sort = "created"})`

Ha yes, exposing `zk.list` makes sense. I guess `zk.tag.list` as well. Any other command is missing for you in the LSP server?

**Reply by mhanberg** [reply]
I think those are currently the only two.

I assume i would be able to do `zk.list({["link-to"] = vim.fn.expand("%")})` to get back links, etc

**Reply by mickael-menu** [reply]
Yes, that should work.

**Reply by mickael-menu** [reply]
@mhanberg I implemented the `zk.list` and `zk.tag.list` LSP commands in https://github.com/mickael-menu/zk/pull/114

[Take a look at the documentation for the available options](https://github.com/mickael-menu/zk/blob/main/docs/editors-integration.md#zklist), the naming doesn't 100% match the CLI flags. And in particular, you need to provide a `select` option to explicitly give the note fields you want in the output.

**Reply by mhanberg** [reply]
Nice!

I was able to update my config to this, which no longer has to shell out to the cli tool.

This is probably not the design i'll stick to, but is basically a direct port from cli calls to lsp calls.

```lua
local nnoremap = require("motch.utils").nnoremap
local _ = require("underscore")

vim.cmd([[command! ZkDaily :term zk daily]])

local colors = require("ansicolors")

local zk = {}

zk.lsp = {}

local command = function(command, ...)
  return vim.lsp.buf_request_sync(0, "workspace/executeCommand", {
    command = command,
    arguments = { vim.api.nvim_buf_get_name(0), ... },
  })[1]["result"]
end

zk.lsp.list = function(args)
  return command("zk.list", args)
end

zk.lsp.tag_list = function(args)
  return command("zk.tag.list", args)
end

zk.note_picker = function(...)
  local notes = zk.lsp.list(...) or {}

  notes = _.map(notes, function(n)
    local tags = _.join(
      _.map(n.tags or {}, function(t)
        return "#" .. t
      end),
      " "
    )

    return colors.noReset("%{bright}%{yellow}")
      .. n.title
      .. colors("%{reset}")
      .. " "
      .. ":"
      .. " "
      .. tags
      .. " "
      .. ":"
      .. " "
      .. n.absPath
  end)

  local actions = {
    ["enter"] = function(selected)
      vim.fn.execute("edit " .. selected)
    end,
    ["ctrl-v"] = function(selected)
      vim.fn.execute("vsplit " .. selected)
    end,
    ["ctrl-x"] = function(selected)
      vim.fn.execute("split " .. selected)
    end,
    ["ctrl-n"] = function(_, query)
      require("lspconfig/configs").zk.new({ title = query })
    end,
  }

  FZF({
    source = notes,
    sinklist = function(selected)
      local query = selected[1]
      local action = actions[selected[2]]
      local parts = vim.fn.split(selected[3], ":")

      action(parts[#parts], query)
    end,
    options = {
      "--expect",
      "enter,ctrl-v,ctrl-x,ctrl-n",
      "--header",
      colors("%{blue}CTRL-N: create a note with the query as title"),
      "--print-query",
      "--ansi",
      "--delimiter",
      ":",
      "--preview",
      "bat {-1}",
      "--nth",
      "1..2",
      "--prompt",
      "Notes> ",
    },
    window = { width = 0.9, height = 0.6, yoffset = 0, highlight = "Normal" },
  })
end

zk.list = function()
  zk.note_picker({ select = { "title", "tags", "absPath" }, sort = { "created" } })
end

zk.tags = function()
  local tags = zk.lsp.tag_list()

  tags = _.map(tags, function(t)
    return t.name
  end)

  FZF({
    source = tags,
    sink = function(tag)
      zk.note_picker({ select = { "title", "tags", "absPath" }, sort = { "created" }, tags = { tag } })
    end,
    options = { "--ansi", "--prompt", "Tags> " },
    window = { width = 0.9, height = 0.6, yoffset = 0, highlight = "Normal" },
  })
end

zk.backlinks = function()
  zk.note_picker({ select = { "title", "tags", "absPath" }, sort = { "created" }, linkTo = {vim.fn.expand("%")} })
end

zk.links = function()
  zk.note_picker({ select = { "title", "tags", "absPath" }, sort = { "created" }, linkedBy = {vim.fn.expand("%")} })
end

_G.motch.zk = zk

local dir = require("lspconfig").util.root_pattern(".zk")(vim.fn.getcwd())
if type(dir) == "string" then
  nnoremap("<C-p>", ":lua motch.zk.list()<cr>")
  nnoremap("<space>zt", ":lua motch.zk.tags()<cr>")
  nnoremap("<space>zb", ":lua motch.zk.backlinks()<cr>")
  nnoremap("<space>zl", ":lua motch.zk.links()<cr>")
  nnoremap("<space>zd", ":ZkDaily<cr>")
end
```

**Reply by mhanberg** [reply]
Probably not much better, but I switched my implementation to this

```lua
local colors = require("ansicolors")

local zk = {}

zk.lsp = setmetatable({}, {
  __index = function(_, key)
    return function(args, callback)
      vim.lsp.buf_request(0, "workspace/executeCommand", {
        command = "zk." .. key,
        arguments = { vim.api.nvim_buf_get_name(0), args },
      }, callback)
    end
  end,
})

local prepend_hash = function(str)
  return "#" .. str
end

local default_opts = function(opts)
  return vim.tbl_extend("force", { select = { "title", "tags", "absPath" }, sort = { "created" } }, opts or {})
end

zk.note_picker = function(...)
  zk.lsp.list(default_opts(...), function(err, result)
    if err then
      vim.api.nvim_err_write("Error fetching notes from zk: " .. vim.inspect(err))
      return
    end

    local notes = vim.tbl_map(function(n)
      local sep = " "
      local tags = vim.fn.join(vim.tbl_map(prepend_hash, n.tags or {}), sep)

      return colors.noReset("%{bright}%{yellow}")
        .. n.title
        .. colors("%{reset}")
        .. " "
        .. ":"
        .. " "
        .. tags
        .. " "
        .. ":"
        .. " "
        .. n.absPath
    end, result)

    local actions = {
      ["enter"] = function(selected)
        vim.fn.execute("edit " .. selected)
      end,
      ["ctrl-v"] = function(selected)
        vim.fn.execute("vsplit " .. selected)
      end,
      ["ctrl-x"] = function(selected)
        vim.fn.execute("split " .. selected)
      end,
      ["ctrl-n"] = function(_, query)
        require("lspconfig/configs").zk.new({ title = query })
      end,
    }

    FZF({
      source = notes,
      sinklist = function(selected)
        local query = selected[1]
        local action = actions[selected[2]]
        local parts = vim.fn.split(selected[3], ":")

        action(parts[#parts], query)
      end,
      options = {
        "--expect",
        "enter,ctrl-v,ctrl-x,ctrl-n",
        "--header",
        colors("%{blue}CTRL-N: create a note with the query as title"),
        "--print-query",
        "--ansi",
        "--delimiter",
        ":",
        "--preview",
        "bat {-1}",
        "--nth",
        "1..2",
        "--prompt",
        "Notes> ",
      },
      window = { width = 0.9, height = 0.6, yoffset = 0, highlight = "Normal" },
    })
  end)
end

zk.list = zk.note_picker

zk.tags = function()
  zk.lsp["tag.list"](nil, function(err, result)
    if err then
      vim.api.nvim_err_write("Error fetching tags from zk: " .. vim.inspect(err))
      return
    end

    local tags = vim.tbl_map(function(t)
      return t.name
    end, result)

    FZF({
      source = tags,
      sink = function(tag)
        zk.note_picker({ tags = { tag } })
      end,
      options = { "--ansi", "--prompt", "Tags> " },
      window = { width = 0.9, height = 0.6, yoffset = 0, highlight = "Normal" },
    })
  end)
end

zk.backlinks = function()
  zk.note_picker({ linkTo = { vim.fn.expand("%") } })
end

zk.links = function()
  zk.note_picker({ linkedBy = { vim.fn.expand("%") } })
end

_G.motch.zk = zk
```

**Comment by mhanberg** [comment]
Here is my cleaned up integration of the current zk lsp commands (new, list, tags.list) and fzf. I realized it made more sense to separate a "zk" file from a "notes" file that does the actual fzf integration.

https://github.com/mhanberg/.dotfiles/blob/main/config/nvim/lua/motch/zk.lua
https://github.com/mhanberg/.dotfiles/blob/main/config/nvim/plugin/notes.lua#L108..L119

**Reply by mickael-menu** [reply]
Nice, thanks! Last time I tried using your config I was a bit confused by this `motch` dependency. Is it just a local package you made, what is "motch" referring to? Please bear with me, as I'm not very familiar with nvim Lua.

Would you mind sharing how to use your config with regular vim-plug for the dependency management?

**Reply by mhanberg** [reply]
Haha!

`motch` is just a silly nickname my friends bestowed upon me, that I am using here as a "namespace" for my local lua modules. The code is all local to my dotfiles, so you can't really use it as a plugin.

the code located in `/lua/**/*.lua` are just lua modules that one can use by requiring them in other code `require("motch.zk")` will refer to `/lua/motch/zk.lua`.

the code located in `plugin/*.lua` are lua files that will be sourced automatically when neovim is booting up. I believe these are run after `init.lua` is loaded.

I can extract it into a plugin real quick to make it easier to try without needing random stuff from my own config.

**Reply by mickael-menu** [reply]
> `motch` is just a silly nickname my friends bestowed upon me, that I am using here as a "namespace" for my local lua modules. The code is all local to my dotfiles, so you can't really use it as a plugin.

That makes sense now 😁 

> I can extract it into a plugin real quick to make it easier to try without needing random stuff from my own config.

It would be appreciated! And could be a nice way to really quickstart https://github.com/mickael-menu/zk-nvim

**Reply by mhanberg** [reply]
Let me know if you're able to get this installed and running: https://github.com/mhanberg/zk.nvim

**Reply by mickael-menu** [reply]
@mhanberg Yes it's working great, thanks, I'll explore this!

**Comment by mhanberg** [comment]
Pushed some updates to my zk.nvim plugin and included some links to examples in my dotfiles how i'm using it for some more bespoke things.

https://github.com/mhanberg/zk.nvim/commit/2ac33167a747a336a6b2a29ca36ac958c9495baf

**Reply by mhanberg** [reply]
I'd like to figure out a way to visualize the `zk graph` in neovim, but not sure where to start haha.

**Reply by mickael-menu** [reply]
By the way there has been some discussion and work done on https://github.com/mickael-menu/zk-nvim/pull/5, it's using Telescope but I think we could have both fzf and telescope support as an opt-in basis. More importantly, each fuzzy finder plugin could share a common wrapper around the LSP server.

> I'd like to figure out a way to visualize the `zk graph` in neovim, but not sure where to start haha.

Yeah I've been thinking about this too but couldn't find a satisfying solution for a TUI. I opened a dedicated discussion to talk about this: https://github.com/mickael-menu/zk/discussions/125

**Reply by mhanberg** [reply]
> More importantly, each fuzzy finder plugin could share a common wrapper around the LSP server.

Oh for sure, I knew there was some WIP on that and didn't want to submit anything until they had time to get started.

**Reply by mickael-menu** [reply]
Sounds good! Feel free to chime in in the discussions
